<a href="https://colab.research.google.com/github/zi-onion/24-2-ESAA/blob/main/09_27_%ED%85%8D%EC%8A%A4%ED%8A%B8%EB%B6%84%EB%A5%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 04 텍스트 분류 실습 - 20 뉴스그룹 분류
fetch_20newsgroups(): 뉴스그룹 예제 데이터

### **텍스트 정규화**


In [28]:
from sklearn.datasets import fetch_20newsgroups

news_data = fetch_20newsgroups(subset='all', random_state=156)

In [4]:
print(news_data.keys())

dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])


In [5]:
import pandas as pd

print('target 클래스의 값과 분포도 \n', pd.Series(news_data.target).value_counts().sort_index())
print('target 클래스의 이름들 \n', news_data.target_names)

target 클래스의 값과 분포도 
 0     799
1     973
2     985
3     982
4     963
5     988
6     975
7     990
8     996
9     994
10    999
11    991
12    984
13    990
14    987
15    997
16    910
17    940
18    775
19    628
Name: count, dtype: int64
target 클래스의 이름들 
 ['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']


->target은 0~19까지 구성

In [6]:
# 개별 데이터 확인
print(news_data.data[0])

From: egreen@east.sun.com (Ed Green - Pixel Cruncher)
Subject: Re: Observation re: helmets
Organization: Sun Microsystems, RTP, NC
Lines: 21
Distribution: world
Reply-To: egreen@east.sun.com
NNTP-Posting-Host: laser.east.sun.com

In article 211353@mavenry.altcit.eskimo.com, maven@mavenry.altcit.eskimo.com (Norman Hamer) writes:
> 
> The question for the day is re: passenger helmets, if you don't know for 
>certain who's gonna ride with you (like say you meet them at a .... church 
>meeting, yeah, that's the ticket)... What are some guidelines? Should I just 
>pick up another shoei in my size to have a backup helmet (XL), or should I 
>maybe get an inexpensive one of a smaller size to accomodate my likely 
>passenger? 

If your primary concern is protecting the passenger in the event of a
crash, have him or her fitted for a helmet that is their size.  If your
primary concern is complying with stupid helmet laws, carry a real big
spare (you can put a big or small head in a big helmet, bu

-> 뉴스그룹 기사, 뉴스그룹 제목, 작성자, 소속, 이메일 등 정보 가짐



In [29]:
# 뉴스 내용 제외하고 삭제
from sklearn.datasets import fetch_20newsgroups

#subset='train'으로 학습용 데이터만 추출, remove=('headers','footers','quotes')로 내용만 추출
train_news = fetch_20newsgroups(subset='train',remove=('headers','footers','quotes'),
                                random_state=156)

X_train = train_news.data
y_train = train_news.target

#subset='test'으로 학습용 데이터만 추출, remove=('headers','footers','quotes')로 내용만 추출
test_news = fetch_20newsgroups(subset='test',remove=('headers','footers','quotes'),
                                random_state=156)

X_test = test_news.data
y_test = test_news.target

print('학습 데이터 크기{0}, 테스트 데이터 크기 {1}'.format(len(train_news.data),
                                            len(test_news.data)))

학습 데이터 크기11314, 테스트 데이터 크기 7532


### **피처 벡터화 변환과 머신러닝 모델 학습/예측/평가**

테스트 데이터의 CounterVerctorizer 적용시, 피처 개수가 같아야하기 때문에 학습 데이터를 이용해 fit된 객체를 이용해서 변환해야 함. 또한, fit_transform() 사용도 불가함.




CountVectorizer 기반 예측

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

# Count Vectorization으로 피처 백터화 변환 수행
cnt_vect = CountVectorizer()
cnt_vect.fit(X_train)
X_train_cnt_vect = cnt_vect.transform(X_train)

# 학습 데이터로 fit()된 Countvextorizer를 이용해 테스트 데이터를 피터 벡터화 변환 수행
X_test_cnt_vect = cnt_vect.transform(X_test)

print('학습 데이터 테스트의 CountVectorizer Shape:', X_train_cnt_vect.shape)

학습 데이터 테스트의 CountVectorizer Shape: (11314, 101631)


In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')

# LogisticRegression을 이용하여 학습/예측/평가 수행
lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train_cnt_vect, y_train)
pred = lr_clf.predict(X_test_cnt_vect)
print('CountVectorized Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test,pred)))


CountVectorized Logistic Regression의 예측 정확도는 0.617


-> 예측 정확도가 약 0.616.

TF-IDF 기반 예측

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF 벡터화를 적용해 학습 데이터 세트와 테스트 데이터 세트 변환.
tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

# LogisticRegression을 이용해 학습/예측/평가 수행.
lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train_tfidf_vect, y_train)
pred = lr_clf.predict(X_test_tfidf_vect)
print('TF-IDF Logistic Regression의 예측 정확도는 {0:.3f}'.format(accuracy_score(y_test,pred)))

TF-IDF Logistic Regression의 예측 정확도는 0.678


예측 정확도는 0.678로 TF-IDF가 단순 카운트 기반보다 더 높은 예측 정확도를 제공함.

-> 텍스트가 많은 문서의 경우 카운트 벡터화보다 TF-IDF가 더 유용.


텍스트 분석에서 머신러닝 모델 성능 향상시는 방법
* 최적의 ML알고리즘 선택
* 최상의 피처 전처리 수행



In [18]:
# stop words 필터링을 추가, ngram을 (1,2)로 변경하여 피처 벡터화
tfidf_vect = TfidfVectorizer(stop_words='english', ngram_range=(1,2), max_df=300)
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train_tfidf_vect,y_train)
pred = lr_clf.predict(X_test_tfidf_vect)
print('TF-IDF Vectorized Logistic Regression의 예측 정확도는 {0:.3f}'.format(
    accuracy_score(y_test,pred)))

TF-IDF Vectorized Logistic Regression의 예측 정확도는 0.690


GridSearchCV를 이용한 최적화 수행

In [19]:
from sklearn.model_selection import GridSearchCV

# 최적 C값 도출 튜닝 수행. CV는 3 폴드 세트로 설정
params = {'C':[0.01, 0.1, 1, 5, 10]}
grid_cv_lr = GridSearchCV(lr_clf, param_grid=params, cv=3,scoring='accuracy',verbose=1)
grid_cv_lr.fit(X_train_tfidf_vect, y_train)
print('Logistic Regression best C parameter :', grid_cv_lr.best_params_)

# 최적 C의 값으로 학습된 grid_cv로 예측 및 정확도 평가
pred = grid_cv_lr.predict(X_test_tfidf_vect)
print('TF-IDF Vectorized Logistic Regression의 예측 정확도는 {0:.3f}'.format(
    accuracy_score(y_test,pred)))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
Logistic Regression best C parameter : {'C': 10}
TF-IDF Vectorized Logistic Regression의 예측 정확도는 0.704


-> C가 10일 때 GridSearchCV의 교차 검증 테스트 세트에서 가장 좋은 예측 성능을 나타내며, 그때 성능은 0.704롤 조금 향상됨.

### **사이킷런 파이프라인(Pipeline) 사용 및 GridSearchCV와의 결합**

* Pipline 클래스: 피처 벡터화와 ML 알고리즘 학습/예측 한번에 진행 가능하게 함.
직관적. 수행 시간 절약 가능

In [22]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([('tfidf_vect', TfidfVectorizer(stop_words='english')),
                     ('lr_clf',LogisticRegression(random_state=156))])

-> TfidfVectorizer객체를 tfidC_vect라는 객체 변수명으로
, LogisticRegression객체를 lr_clf라
는 객체 변수명으로 생성한 뒤 이 두 개의 객체를 파이프라인으로 연결하는 Pipine이 객체 pipeline을
생성

In [24]:
from sklearn.pipeline import Pipeline

# TfidfVectorizer 객체를 tfidf_vect로, LogisticRegression 객체를 lr_clf로 생성
pipeline = Pipeline([
     ('tfidf_vect', TfidfVectorizer(stop_words='english', ngram_range=(1,2),max_df=300)),
                     ('lr_clf',LogisticRegression(solver='liblinear', C=10))
                     ])

In [25]:
# 별도의 TfidfVectorizer 객체의 fit(), transform()과 LogisticRegression의 fit(),predict()가 필요 없음
# pipeline의 fit(),predict()만으로 한꺼번에 피처 벡터화와 ML학습/예측이 가능.

pipeline.fit(X_train, y_train)
pred = pipeline.predict(X_test)
print('Pipeline을 통한 Logistic Regression의 예측 정확도는 {0:.3f}'.format(
    accuracy_score(y_test,pred)
))

Pipeline을 통한 Logistic Regression의 예측 정확도는 0.704


In [32]:
# GridSearchCV에 Pipeline을 입력하여 TfidfVectorizer와 LogisticRegression의 하이퍼 파라미터를 함께 최적화
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
     ('tfidf_vect', TfidfVectorizer(stop_words='english')),
                     ('lr_clf',LogisticRegression())
                     ])

# Pipeline에 기술된 각각의 객체 변수에 언더바(_) 2개를 연달아 붙여 GridsearchCV에 사용될 파라미터/하이퍼 파라미터 이름과 값 설정
params = { 'tfidf_vect__ngram_range': [(1,1),(1,2),(1,3)],
           'tfidf_vect__max_df': [100,300,700],
            'lr_clf__C':[1,5,10]
           }

# GridSearchCV의 생성자에 Estimator가 아닌 Pipeline 객체 입력
grid_cv_pipe = GridSearchCV(pipeline, param_grid=params, cv=3, scoring='accuracy', verbose=1)
grid_cv_pipe.fit(X_train, y_train)
print(grid_cv_pipe.best_params_, grid_cv_pipe.best_score_)

pred = grid_cv_pipe.predict(X_test)
print('Pipeline을 통한 LogisticRegression의 에측 정확도는{0:.3f}'.format(
    accuracy_score(y_test,pred)))

Fitting 3 folds for each of 27 candidates, totalling 81 fits


KeyboardInterrupt: 

-> TfidfVectorizer 객체의 max_df 파라미터가 700, ngramjrange 파라미터가 (1,2)로 피처 벡터화된 데이터 세트에 LogisticRegression의 C 하이퍼 파라미터에
10을 적용해 예측 분류를 수행할 때 가장
좋은 검증 세트 성능 수치 0.702가 도출됨.

## 05 감성 분석

### **감성 분석 소개**
* 감성 분석(Sentiment Analysis): 문서의 주관적인 감성/의견/감정/기분 등을 파악하기 위한 방법

-> 텍스트가 나타내는 주관적인 단어와 문맥으로 감성 수치를 계산하고 긍정/부정 감성을 결정.


  - 지도학습은 학습 데이터와 타깃 레이블 값을 기반으로 감성 분석 학습을 수행한 뒤 이를 기반으로 다른 데이터의 감성 분석을 예측하는 방법으로, 일반적인 텍스트 기반의 분류와 거의 동일
  - 비지도학습은 ‘Lexicon’이라는 일종의 감성 어휘 사전을 이용. Lexicon은 감성 분석을 위한 용어와 문맥에 대한 다양한 정보를 가지고 있으며, 이를 이용해 문서의 긍정적, 부정적 감성 여부를 판단.

### **지도학습 기반 감성 분석 실습 - IMDB 영화평**

In [35]:
import pandas as pd
review_df = pd.read_csv('/content/labeledTrainData.tsv', header=0, sep="\t", quoting=3)
review_df.head(3)

id  sentiment                                             review
0  "5814_8"          1  "With all this stuff going down at the moment ...
1  "2381_9"          1  "\"The Classic War of the Worlds\" by Timothy ...
2  "7759_3"          0  "The film starts with a manager (Nicholas Bell...

- id： 각데이터의 id
- sentiment： 영화평(review)의 Sentiment 결과 값(Target Label). 1
은 긍정적 평가, 0은 부정적 평가를 의미.
- review： 영화평의 텍스트

In [36]:
# 텍스트 값 확인
print(review_df['review'][0])

"With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.<br /><br />Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.<br /><br />The actual feature film bit when it finally sta

In [37]:
# 텍스트 전처리
import re

# <br\> html 태그는 replace 함수로 공백으로 변환
review_df['review'] = review_df['review'].str.replace('<br/>',' ')

# 파이썬의 정규 표현식 모듈인 re를 이용해 영어 문자열이 아닌 문자는 모두 공백으로 변환
review_df['review'] = review_df['review'].apply( lambda x : re.sub("[Aa-zA—Z]", " ", x) )

In [38]:
from sklearn.model_selection import train_test_split

class_df = review_df['sentiment'] # 결정값 클래스 따로 추출
feature_df = review_df.drop(['id','sentiment'],axis=1,inplace=False)

X_train,X_test,y_train,y_test = train_test_split(feature_df, class_df, test_size = 0.3,
                                                 random_state=156)
X_train.shape, X_test.shape

((17500, 1), (7500, 1))

In [40]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

# 스톱워드는 English, ngram은 (1,2)로 설정해 CountVectorization 수행
# LogisticRegression의 C는 10으로 설정
pipeline = Pipeline([
    ('cnt_vect', CountVectorizer(stop_words='english', ngram_range=(1,2))),
    ('lr_clf',LogisticRegression(solver='liblinear',C=10))])

# Pipeline 객체를 이용해 fit(), predict()로 학습/예측 수행. predict_proba() roc_auc 때문에 수행.
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:,-1]

print('예측 정확도는 {0:.4f}, ROC-AUC는 {1:.4f}'.format(accuracy_score(y_test,pred),
                                                 roc_auc_score(y_test, pred_probs)))


예측 정확도는 0.5688, ROC-AUC는 0.6008


In [41]:
# 스톱워드는 English, ngram은 (1,2)로 설정해 TF-IDF 벡터화 수행.
# LogisticRegression의 C는 10으로 설정
pipeline = Pipeline([
    ('tfidf_vect', TfidfVectorizer(stop_words='english', ngram_range=(1,2))),
    ('lr_clf',LogisticRegression(solver='liblinear',C=10))])

pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:,-1]

print('예측 정확도는 {0:.4f}, ROC-AUC는 {1:.4f}'.format(accuracy_score(y_test,pred),
                                                 roc_auc_score(y_test, pred_probs)))


예측 정확도는 0.5725, ROC-AUC는 0.6047


-> CountVectorizer보다 TF-IDF가 조금 나아짐

### **비지도학습 기반 감성 분석 소개**

:Lexicon 기반. Lexicon은 일반적으로 어휘집을 의미하지만 여기서는 주로 감성만을 분석하기 위해 지원하는 감성어휘 사전

ex) NLTK 패키지

NLTK 패키지의 WordNet: 시맨틱 분석을 제공하는 어휘 사전
  - 시맨틱(semantic):‘문맥상 의미’


[감성사전 예]
- SentiWordNet： NLTK 패키지의 WordNet과 유사하게 감성 단어 전용의 WordNet을 구현한 것
- VADER： 주로 소셜 미디어의 텍스트에 대한 감성 분석을 제공하기 위한 패키지
- Pattern： 예측 성능 측면에서 가장 주목받는 패키지

### **SentiWordNet 을 이용한 감성 분석**

**WordNet SynsetZF SentiWordNet SentiSynset 클래스의 이해**


In [42]:
import nltk
nltk.download('all')


[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_eng.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_rus to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |  

True

In [43]:
from nltk.corpus import wordnet as wn

term = 'present'

# 'present'라는 단어로 wordnet의 synsets 생성.
synsets = wn.synsets(term)
print('syntets()반환 type:',type(synsets))
print('syntets()반환 값 개수:', len(synsets))
print('syntets()반환 값:',synsets)

syntets()반환 type: <class 'list'>
syntets()반환 값 개수: 18
syntets()반환 값: [Synset('present.n.01'), Synset('present.n.02'), Synset('present.n.03'), Synset('show.v.01'), Synset('present.v.02'), Synset('stage.v.01'), Synset('present.v.04'), Synset('present.v.05'), Synset('award.v.01'), Synset('give.v.08'), Synset('deliver.v.01'), Synset('introduce.v.01'), Synset('portray.v.04'), Synset('confront.v.03'), Synset('present.v.12'), Synset('salute.v.06'), Synset('present.a.01'), Synset('present.a.02')]


In [47]:
for synset in synsets:
  print('##### Synset name: ', synset.name(), '#####')
  print('POS ：', synset.lexname())
  print('Definition：',synset.definition())
  print('Lemmas:', synset.lemma_names())

##### Synset name:  present.n.01 #####
POS ： noun.time
Definition： the period of time that is happening now; any continuous stretch of time including the moment of speech
Lemmas: ['present', 'nowadays']
##### Synset name:  present.n.02 #####
POS ： noun.possession
Definition： something presented as a gift
Lemmas: ['present']
##### Synset name:  present.n.03 #####
POS ： noun.communication
Definition： a verb tense that expresses actions or states at the time of speaking
Lemmas: ['present', 'present_tense']
##### Synset name:  show.v.01 #####
POS ： verb.perception
Definition： give an exhibition of to an interested audience
Lemmas: ['show', 'demo', 'exhibit', 'present', 'demonstrate']
##### Synset name:  present.v.02 #####
POS ： verb.communication
Definition： bring forward and present to the mind
Lemmas: ['present', 'represent', 'lay_out']
##### Synset name:  stage.v.01 #####
POS ： verb.creation
Definition： perform (a play), especially on a stage
Lemmas: ['stage', 'present', 'represent']
##

-> Synset('present.n.01‘)과 Synset('present.n.02')는 명사지만 서로 다른 의미를 가짐. 각각 현재, 선물의 의미.

=> Synset: 나의 단어가 가질 수 있는 여러 가지 시맨틱 정보를 개별 클래스로 나타낸 것

WordNet은 어떤 어휘와 다른 어휘 간의 관계를 유사도로 나타낼 수 있음

In [49]:
# synset 객체를 단어별로 생성합니다
tree = wn.synset('tree.n.01')
lion = wn.synset('lion.n.01')
tiger = wn.synset('tiger.n.02')
cat = wn.synset('cat.n.01')
dog = wn.synset('dog.n.01')
entities = [tree, lion, tiger, cat, dog]
similarities = []
entity_names = [entity.name().split('.')[0] for entity in entities]

# 단어별 synset을 반복하면서 다른 단어의 synset과 유사도를 측정합니다
for entity in entities:
  similarity = [round(entity.path_similarity(compared_entity), 2)
                for compared_entity in entities]
  similarities.append(similarity)

# 개별 단어별 synset과 다른 단어의 synset과의 유사도를 DataFrame 형태로 저장합니다.
similarity_df = pd.DataFrame(similarities, columns=entity_names, index=entity_names)
similarity_df

tree  lion  tiger   cat   dog
tree   1.00  0.07   0.07  0.08  0.12
lion   0.07  1.00   0.33  0.25  0.17
tiger  0.07  0.33   1.00  0.25  0.17
cat    0.08  0.25   0.25  1.00  0.20
dog    0.12  0.17   0.17  0.20  1.00

-> lion은 tree와의 유사도가 0.07로 가장 작고, tiger와는 유사도가 0.33으로 가장 큼.



In [51]:
import nltk
from nltk.corpus import sentiwordnet as swn

senti_synsets = list(swn.senti_synsets('slow'))
print('senti_synsets() 반환 type ：', type(senti_synsets))
print('senti_synsets() 반환 값 개수:', len(senti_synsets))
print('senti_synsets() 반환 값:', senti_synsets)

senti_synsets() 반환 type ： <class 'list'>
senti_synsets() 반환 값 개수: 11
senti_synsets() 반환 값: [SentiSynset('decelerate.v.01'), SentiSynset('slow.v.02'), SentiSynset('slow.v.03'), SentiSynset('slow.a.01'), SentiSynset('slow.a.02'), SentiSynset('dense.s.04'), SentiSynset('slow.a.04'), SentiSynset('boring.s.01'), SentiSynset('dull.s.08'), SentiSynset('slowly.r.01'), SentiSynset('behind.r.03')]


SentiSynset 객체: 단어의 감성을 나타내는 감성 지수와 객관성을 나타내는 객관성 지
수를 가지고 있음.

In [53]:
import nltk
from nltk.corpus import sentiwordnet as swn

father = swn.senti_synset('father.n.01')
print('father 긍정감성 지수: ', father.pos_score())
print('father 부정감성 지수: ',father.neg_score ())
print('father 객관성 지수: ', father.obj_score())
print('\n')
fabulous = swn.senti_synset('fabulous.a.01')
print('fabulous 긍정감성 지수:', fabulous.pos_score())
print('fabulous 부정감성 지수:', fabulous.neg_score())

father 긍정감성 지수:  0.0
father 부정감성 지수:  0.0
father 객관성 지수:  1.0


fabulous 긍정감성 지수: 0.875
fabulous 부정감성 지수: 0.125


-> father는 객관적인 단어로 객관성 지수가 1.0이고 긍정 감성/부정 감성 지수 모두 0

**SentiWordNet을 이용한 영화 감상평 감성 분석**

품사 태깅(POS Tagging)

In [54]:
from nltk.corpus import wordnet as wn

# 간단한 NTLK PennTreebank Tag를 기반으로 WordNet 기반의 품사 Tag로 변환
def penn_to_wn(tag):
  if tag.startswith('3'):
    return wn.ADJ
  elif tag.startswith('N'):
    return wn.NOUN
  elif tag.startswith('R'):
    return wn.ADV
  elif tag.startswith('V'):
    return wn.VERB


문서를 문장 —> 단어 토큰 一> 품사 태깅 후에 SentiSynset 클래스를 생성하고 Polarity Score를 합산하는 함수를 생성

In [58]:
from nltk.stem import WordNetLemmatizer
from nltk.corpus import sentiwordnet as swn
from nltk import sent_tokenize, word_tokenize, pos_tag

def swn_polarity(text):
  # 감성 지수 초기화
  sentiment = 0.0
  tokens_count = 0

  lemmatizer = WordNetLemmatizer()
  raw_sentences = sent_tokenize(text)
  # 분해된 문장별로 단어 토큰 -> 품사 태깅 후에 SentiSynset 생성 -> 감성 지수 합산
  for raw_sentence in raw_sentences:
    # NTLK 기반의 품사 태깅 문장 추출
    tagged_sentence = pos_tag(word_tokenize(raw_sentence))
    for word, tag in tagged_sentence:

      # WordNet 기반 품사 태깅과 어근 추출
      wn_tag = penn_to_wn(tag)
      if wn_tag not in (wn.NOUN, wn.ADJ, wn.ADV):
        continue
      lemma = lemmatizer.lemmatize(word, pos=wn_tag)
      if not lemma:
        continue
      # 어근을 추출한 단어와 WordNet 기반 품사 태깅을 입력해 Synset 객체를 생성
      synsets = wn.synsets(lemma, pos=wn_tag)
      if not synsets:
        continue
      # sentiwordnet의 감성 단어 분석으로 감성 synset 추출
      # 모든 단어에 대해 긍정 감성 지수는 +로 부정 감성 지수는 -로 합산해 감성 지수 계산
      synset = synsets[0]
      swn_synset = swn.senti_synset(synset.name())
      sentiment += (swn_synset.pos_score() - swn_synset.neg_score())
      tokens_count = 1

  if not tokens_count:
    return 0

  # 총 score가 0 이상일 경우 긍정(Positive) 1, 그렇지 않을 경우 부정(Negative) 0 반환
  if sentiment >= 0 :
    return 1

  return 0

swn_polarity(text) 함수를 IMDB 감상평의 개별 문서에 적용해 긍정 및 부정 감성을
예측

In [60]:
review_df['preds'] = review_df['review'].apply( lambda x : swn_polarity(x) )
y_target = review_df['sentiment'].values
preds = review_df['preds'].values

In [64]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score
from sklearn.metrics import recall_score, f1_score, roc_auc_score
import numpy as np

print(confusion_matrix(y_target, preds))
print('정확도: ', np.round(accuracy_score(y_target, preds), 4))
print("정밀도: ", np.round(precision_score(y_target, preds), 4))
print("재현율: ", np.round(recall_score(y_target, preds), 4))

[[7119 5381]
 [7088 5412]]
정확도:  0.5012
정밀도:  0.5014
재현율:  0.433


### **VADER 를이용한 감성분석**

VADER: 소셜 미디어의 감성 분석 용도로
만들어진 룰 기반의 Lexicon

In [67]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

senti_analyzer = SentimentIntensityAnalyzer()
senti_scores = senti_analyzer.polarity_scores(review_df['review'][0])
print(senti_scores)

{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


-> 'neg'는 부정 감성 지수, ‘neu’는 중립적인 감성 지수, ‘pos’는 긍정 감성 지수
, 그리고 compound는 neg, neu, pos score를 적절히 조합해 —1
에서 1 사이의 감성 지수를 표현한 값


보통 compound score가 0.1 이상이면 긍정 감성, 그 이하이면 부정감성으로 결정

In [70]:
def vader_polarity(review, threshold=0.1):
  analyzer = SentimentIntensityAnalyzer()
  scores = analyzer.polarity_scores(review)

  # compound 값에 기반해 threshold 입력값보다 크면 1, 그렇지 않으면 0을 반환
  agg_score = scores['compound']
  final_sentiment = 1 if agg_score >= threshold else 0
  return final_sentiment

# apply lambda 식을 이용해 레코드별로 vader_polarity()를 수행하고 결과를 'vader_preds'에 저장
review_df['vader_preds'] = review_df['review'].apply(lambda x : vader_polarity(x, 0.1))
y_target = review_df['sentiment'].values
vader_preds = review_df['vader_preds'].values

print(confusion_matrix(y_target, vader_preds))
print("정확도:", np.round(accuracy_score(y_target, vader_preds),4))
print("정밀도:", np.round(precision_score(y_target , vader_preds),4))
print("재현율:", np.round(recall_score(y_target, vader_preds),4))

[[11685   815]
 [11864   636]]
정확도: 0.4928
정밀도: 0.4383
재현율: 0.0509


-> 재현율이 SentiWordNet보다 향상됨.